### PR TITLE
Add tests for active path discovery

### PR DESCRIPTION
**Pull request by danderson**
---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/601)

### CODE REVIEW

**Comment by danderson** [comment]
@bradfitz ready for review. New tests explicitly verify that magicsock disco is finding direct paths between peers in various simple-ish NAT traversal cases. It also includes a bunch of machinery that should, hopefully, make writing tests for more elaborate conditions easier in future (e.g. I'd really like a test for downgrade/redisco).

File: tstest/natlab/natlab.go [hunk 1]
```diff
@@ -31,6 +31,18 @@ import (
 
 var traceOn, _ = strconv.ParseBool(os.Getenv("NATLAB_TRACE"))
 
+// TODO(danderson): this makes it impossible to parallelize natlab
```
Comment (bradfitz) [inline tstest/natlab/natlab.go:34]: This also probably violates the sync.WaitGroup rules about all Adds having to happen before a Wait. If an Add and Wait happen concurrently, things explode: https://golang.org/pkg/sync/#WaitGroup.Add

File: wgengine/magicsock/magicsock.go [hunk 2]
```diff
@@ -55,6 +55,44 @@ import (
...
+	// full payload.
+	logDerpVerbose, _ = strconv.ParseBool(os.Getenv("TS_DEBUG_DERP"))
+	// debugReSTUNStopOnIdle unconditionally enables the "shut down
+	// STUN if magicsock is idle" behavior that normally only triggers
+	// on mobile devices, lowers the shutdown interval, and logs more
+	// verbosely about idle measurements.
+	debugReSTUNStopOnIdle, _ = strconv.ParseBool(os.Getenv("TS_DEBUG_RESTUN_STOP_ON_IDLE"))
+)
+
+// inTest binds magicsock to 127.0.0.1 instead of its usual 0.0.0.0,
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:84]: say what it reports, not what its caller does with it.

File: wgengine/magicsock/magicsock_test.go [hunk 3]
```diff
@@ -120,6 +123,7 @@ type magicStack struct {
 	tun        *tuntest.ChannelTUN // tuntap device to send/receive packets
 	tsTun      *tstun.TUN          // wrapped tun that implements filtering and wgengine hooks
 	dev        *device.Device      // the wireguard-go Device that connects the previous things
+	tsIP       chan netaddr.IP     // buffered, guaranteed to yield at least 1 value
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock_test.go:121]: what's the value?

File: wgengine/magicsock/magicsock_test.go [hunk 4]
```diff
@@ -190,6 +195,139 @@ func (s *magicStack) Close() {
...
+// AwaitIP waits for magicStack to receive a Tailscale IP address on
+// tsIP, and returns the IP. It's intended for use with magicStacks
+// that have been meshed with meshStacks, to wait for configs to have
+// propagated enough that everyone has a Tailscale IP that should
+// work.
+func (s *magicStack) AwaitIP() netaddr.IP {
+	select {
+	case ip := <-s.tsIP:
+		return ip
+	case <-time.After(2 * time.Second):
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock_test.go:217]: leaking goroutine time here

File: wgengine/magicsock/magicsock_test.go [hunk 5]
```diff
@@ -190,6 +195,139 @@ func (s *magicStack) Close() {
...
+func (s *magicStack) Ping(src, dst netaddr.IP) {
+	pkt := tuntest.Ping(dst.IPAddr().IP, src.IPAddr().IP)
+	s.tun.Outbound <- pkt
+}
+
+func (s *magicStack) AwaitPacket(timeout time.Duration) bool {
+	select {
+	case <-s.tun.Inbound:
+		return true
+	case <-time.After(timeout):
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock_test.go]: likewise

File: wgengine/magicsock/magicsock_test.go [hunk 6]
```diff
@@ -190,6 +195,139 @@ func (s *magicStack) Close() {
...
+	case ip := <-s.tsIP:
+		return ip
+	case <-time.After(2 * time.Second):
+		panic("timed out waiting for magicStack to get an IP")
+	}
+}
+
+func (s *magicStack) Ping(src, dst netaddr.IP) {
+	pkt := tuntest.Ping(dst.IPAddr().IP, src.IPAddr().IP)
+	s.tun.Outbound <- pkt
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock_test.go]: can this block forever during a shutdown? select on a closed channel (if there is one?)

File: wgengine/magicsock/magicsock_test.go [hunk 7]
```diff
@@ -504,6 +643,42 @@ type devices struct {
 	stunIP netaddr.IP
 }
 
+func testActiveDiscovery(t *testing.T, d *devices) {
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock_test.go:761]: comments on this func?

File: wgengine/magicsock/magicsock_test.go [hunk 8]
```diff
@@ -190,6 +195,139 @@ func (s *magicStack) Close() {
...
+	s.conn.UpdateStatus(&sb)
+	return sb.Status()
+}
+
+// AwaitIP waits for magicStack to receive a Tailscale IP address on
+// tsIP, and returns the IP. It's intended for use with magicStacks
+// that have been meshed with meshStacks, to wait for configs to have
+// propagated enough that everyone has a Tailscale IP that should
+// work.
+func (s *magicStack) AwaitIP() netaddr.IP {
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock_test.go:213]: take a *testing.T so you can fail on it?

File: wgengine/magicsock/magicsock_test.go [hunk 9]
```diff
@@ -504,6 +703,126 @@ type devices struct {
...
+
+	cleanup = newPinger(t, logf, m1, m2, m1IP, m2IP)
+	defer cleanup()
+
+	// Everything is now up and running, active discovery should find
+	// a direct path between our peers. Wait for it to switch away
+	// from DERP.
+
+	mustDirect := func(m1, m2 *magicStack) {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock_test.go:806]: I thought you were inheriting from some other context with all this. If this is all you want, why not:

```go
    for deadline := time.Now().Add(5 * time.Second); time.Now().Before(deadline); time.Sleep(100 * time.Millisecond) {
             ...
                return
    }
    t.Errorf("magicsock did not find a direct path from %s to %s", m1, m2)
``

File: wgengine/magicsock/magicsock_test.go [hunk 10]
```diff
@@ -432,67 +560,138 @@ func makeNestable(t *testing.T) (logf logger.Logf, setT func(t *testing.T)) {
...
-				m1:     m1,
-				m1IP:   m1if.V4(),
-				m2:     m2,
-				m2IP:   m2if.V4(),
-				stun:   mstun,
-				stunIP: sif.V4(),
-			}
-			testTwoDevicePing(t, n)
-		})
+	t.Run("facing_firewalls", func(t *testing.T) {
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock_test.go:597]: facing_easy_firewalls ?

File: wgengine/magicsock/magicsock_test.go (hunk 3 above)
Comment (danderson) [reply wgengine/magicsock/magicsock_test.go]: Replaced this with using ipnstate to pull the current Tailscale IP(s) out of the network engine. Also means you can get the IP multiple times without weirdness, which is nice.

File: wgengine/magicsock/magicsock_test.go (hunk 5 above)
Comment (danderson) [reply wgengine/magicsock/magicsock_test.go]: Hm, why does the goroutine leak checker not see these? Test completes in much less than the timeout, so if there's leaking goroutines they should show up, but don't. What's up with that?

File: ipn/ipnstate/ipnstate.go [hunk 11]
```diff
@@ -18,13 +18,15 @@ import (
...
 
+	"inet.af/netaddr"
 	"tailscale.com/tailcfg"
 	"tailscale.com/types/key"
 )
 
 // Status represents the entire state of the IPN network.
 type Status struct {
 	BackendState string
+	IPs          []string // Tailscale IP(s) assigned to this node
```
Comment (bradfitz) [inline ipn/ipnstate/ipnstate.go:29]: I guess string for JSON? Does netaddr.IP not JSON-ify well?

File: ipn/ipnstate/ipnstate.go [hunk 12]
```diff
@@ -109,6 +111,17 @@ func (sb *StatusBuilder) AddUser(id tailcfg.UserID, up tailcfg.UserProfile) {
 	sb.st.User[id] = up
 }
 
+func (sb *StatusBuilder) AddIP(ip netaddr.IP) {
```
Comment (bradfitz) [inline ipn/ipnstate/ipnstate.go:114]: godoc & mention that it's for Tailscale IPs, either in godoc or method name?

File: ipn/ipnstate/ipnstate.go (hunk 11 above)
Comment (danderson) [reply ipn/ipnstate/ipnstate.go:29]: Not sure what I was thinking. netaddr.IP stringifies fine. Fixed.

**Pull request merged**